src/NumericalAlgorithms/LinearSolver/LinearSolver.hpp

```diff
@@ -0,0 +1,279 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <memory>
+#include <optional>
+#include <pup.h>
+#include <pup_stl.h>
+#include <type_traits>
+#include <utility>
+
+#include "NumericalAlgorithms/Convergence/HasConverged.hpp"
+#include "Options/Auto.hpp"
+#include "Options/Options.hpp"
+#include "Parallel/CharmPupable.hpp"
+#include "Parallel/PupStlCpp17.hpp"
+#include "Utilities/FakeVirtual.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/Registration.hpp"
+#include "Utilities/Requires.hpp"
+
+namespace LinearSolver::Serial {
+
+/// Indicates the linear solver uses no preconditioner. It may perform
+/// compile-time optimization for this case.
+struct NoPreconditioner {};
+
+/// Indicates the linear solver uses any other registered linear solver as
+/// preconditioner.
+struct AnyPreconditioner {};
+
+/// Registrars for linear solvers
+namespace Registrars {}
+
+/*!
+ * \brief Base class for serial linear solvers that supports factory-creation
+ * and nested preconditioning.
+ *
+ * To enable support for preconditioning in your derived linear solver class,
+ * pass any type that has a `solve` and a `reset` function as the
+ * `Preconditioner` template parameter. It can also be an abstract
+ * `LinearSolver` type, which means that any other linear solver can be used as
+ * preconditioner. Pass `NoPreconditioner` to disable support for
+ * preconditioning.
+ */
+template <typename LinearSolverRegistrars>
+class LinearSolver : public PUP::able {
+ protected:
+  /// \cond
+  LinearSolver() = default;
+  LinearSolver(const LinearSolver&) = default;
+  LinearSolver(LinearSolver&&) = default;
+  LinearSolver& operator=(const LinearSolver&) = default;
+  LinearSolver& operator=(LinearSolver&&) = default;
+  /// \endcond
+
+ public:
+  ~LinearSolver() override = default;
+
+  /// \cond
+  explicit LinearSolver(CkMigrateMessage* m) noexcept;
+  WRAPPED_PUPable_abstract(LinearSolver);  // NOLINT
+  /// \endcond
+
+  using creatable_classes = Registration::registrants<LinearSolverRegistrars>;
+
+  virtual std::unique_ptr<LinearSolver<LinearSolverRegistrars>> get_clone()
+      const = 0;
+
+  /*!
+   * \brief Solve the linear equation \f$Ax=b\f$ where \f$A\f$ is the
+   * `linear_operator` and \f$b\f$ is the `source`.
+   *
+   * - The (approximate) solution \f$x\f$ is returned in the
+   *   `initial_guess_in_solution_out` buffer, which also serves to provide an
+   *   initial guess for \f$x\f$. Not all solvers take the initial guess into
+   *   account, but expect the buffer is sized correctly.
```

Comment:
I suggest "but expect" -> "but all expect" if, indeed, all solvers to expect the buffer to be sized correctly.